### PR TITLE
Stop preprocessing unicode escapes

### LINKED ISSUE #209: JavaParser and Unicode Escaping
Hi

is it intended that parsing and serializing turns

```
public class Sample {

    public static final String SAMPLE="\u00DCbung";

}
```

into

```
public class Sample {

    public static final String SAMPLE = "Übung";
}
```

### LINKED ISSUE #600: JavaParser interprets Unicode escape sequences in comments
In JavaParser 3.0.0 alpha 10 Unicode escape sequences are interpreted. This results in the following behavior:

```java
JavaParser.parse("class Foo { /* \\u002e */ }").toString()
```
expected result:
```java
class Foo {
    /* \u002e */
}
```

actual result:
```java
class Foo {
    /* . */
}
```

### LINKED ISSUE #801: Remember the original unicode escape sequence
Do we want to store the very original code, including the unicode escape sequences?
@cal101 provided a patch for that

https://github.com/ftomassetti/javaparser/pull/5/files

@matozoid is in a better position than me to comment on it

### LINKED ISSUE #1236: I think there is a bug with unicode string literals
Hi fellas,

I have this code:

```
static String unicode_rightPointingTriangle() {
  return "\u25B8";
}
```
and it seems the JavaParser mangles it into, ultimately, "?".

Has this been tested? I encode all my Java sources as UTF-8 - but it shouldn't matter anyways as it is ecaped as \u, as you see.

I think there really is a bug in the parser/pretty printer.

Greetings,
Stefan

### PR DESCRIPTION
**Pull request by matozoid**
Fixes #1236 
Fixes #801 
Fixes #600
Fixes #209 

@ftomassetti @cal101 @Gottox @digulla @cdietrich - @ptitjes tells us in the original #209 that unicode escape sequences should be gone before the lexer even starts. This allows for unicode escapes in every element of the source code. This is what we currently do, and this is what causes the escape sequences from the source code to be irretrievably lost to the user.

I would like to reverse this decision and stop preprocessing the escapes. This will be the result:
- good: the original escape sequences will be kept and can be printed out again.
- bad: it becomes impossible to use escape sequences in many positions. However, non-ASCII characters are only allowed in comments, string and char literals, and identifiers, so if we make sure they are allowed there this should be a very marginal issue. Intellij doesn't even recognize this case.

I think doing this will support a lot more use cases, and will break few to none.

### CODE REVIEW

**Comment by matozoid** [comment]
The important part of this change is removing `JAVA_UNICODE_ESCAPE=true;`, by the way.

**Comment by ptitjes** [comment]
I guess no one uses escapes inside keywords because keyboard letters... Great idea indeed!

**Comment by ftomassetti** [comment]
I guess that for those very few cases in which they want to process unicode escapes in unexpected places we could have some preprocessing before JavaParser kicks in

File: javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
  * A literal string.
  * <br/><code>"Hello World!"</code>
  * <br/><code>"\"\n"</code>
+ * <br/><code>"\u2122"</code>
  * <br/><code>"™"</code>
  * <br/><code>"💩"</code>
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java:42]: I am not sure what GitHub wants to tell me with this icon here :)

File: javaparser-testing/src/test/java/com/github/javaparser/JavaParserTest.java [hunk 2]
```diff
@@ -207,6 +207,6 @@ public void trailingCodeIsAnError() {
     @Test
     public void trailingWhitespaceIsIgnored() {
         BlockStmt blockStmt = JavaParser.parseBlock("{} // hello");
-        assertEquals("\"}\"   <94>   (line 1,col 2)-(line 1,col 2)", blockStmt.getTokenRange().get().getEnd().toString());
+        assertEquals("\"}\"   <95>   (line 1,col 2)-(line 1,col 2)", blockStmt.getTokenRange().get().getEnd().toString());
```
Comment (ftomassetti) [inline javaparser-testing/src/test/java/com/github/javaparser/JavaParserTest.java]: what this "95" represents here?

File: javaparser-testing/src/test/java/com/github/javaparser/ast/comments/CommentTest.java [hunk 3]
```diff
@@ -58,4 +62,11 @@ public void cannotRemoveCommentNotUsedAnywhere() {
         Comment c = new LineComment("A comment");
         assertFalse(c.remove());
     }
+
+    @Test
+    public void mayContainUnicodeEscapes() {
```
Comment (ftomassetti) [inline javaparser-testing/src/test/java/com/github/javaparser/ast/comments/CommentTest.java]: Maybe "unicodeEscapesArePreservedInComments"

File: javaparser-testing/src/test/java/com/github/javaparser/ast/expr/SimpleNameTest.java [hunk 4]
```diff
@@ -42,4 +43,9 @@ public void identifierMustNotBeNull() {
         new SimpleName(null);
     }
 
+    @Test
+    public void mayContainUnicodeEscapes() {
+        SimpleName name = parseSimpleName("xxx\\u2122xxx");
+        assertEquals("xxx\\u2122xxx", name.asString());
+    }
 }
```
Comment (ftomassetti) [inline javaparser-testing/src/test/java/com/github/javaparser/ast/expr/SimpleNameTest.java]: Newline :D

File: javaparser-testing/src/test/java/com/github/javaparser/ast/expr/StringLiteralExprTest.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import static com.github.javaparser.JavaParser.parseExpression;
+import static org.junit.Assert.*;
+
+public class StringLiteralExprTest {
+    @Test
+    public void unicodeEscapesDoNotGetPreprocessed() {
+        StringLiteralExpr omega = parseExpression("\"xxx\\u03a9xxx\"");
+        assertEquals("xxx\\u03a9xxx", omega.getValue());
+    }
+}
```
Comment (ftomassetti) [inline javaparser-testing/src/test/java/com/github/javaparser/ast/expr/StringLiteralExprTest.java]: Newline :D

**Review by ftomassetti — APPROVED** [review]
The idea is good and should solve a significant issue and the implementation also looks good.

Thumbs up!

File: javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java (hunk 1 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java:42]: The omega or the "emoticon"? ;-)

File: javaparser-testing/src/test/java/com/github/javaparser/JavaParserTest.java (hunk 2 above)
Comment (matozoid) [reply javaparser-testing/src/test/java/com/github/javaparser/JavaParserTest.java]: It the token type. Annoying, but it did remind me to put the new token type in the big category switch.

File: javaparser-testing/src/test/java/com/github/javaparser/ast/comments/CommentTest.java (hunk 3 above)
Comment (matozoid) [reply javaparser-testing/src/test/java/com/github/javaparser/ast/comments/CommentTest.java]: Done

File: javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java:42]: the emoticon :)

File: javaparser-testing/src/test/java/com/github/javaparser/JavaParserTest.java (hunk 2 above)
Comment (ftomassetti) [reply javaparser-testing/src/test/java/com/github/javaparser/JavaParserTest.java]: Maybe we can add a comment about that

File: javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java (hunk 1 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/ast/expr/StringLiteralExpr.java:42]: I knew someone was eventually going to notice ;-)

**Comment by stefan-reich** [comment]
Cool

**Comment by davidsusu** [comment]
IMO preprocessing is the right way. Its only drawback is that it is harder to reproduce the original source. Harder, but not impossible. For a solution to a similar problem see [javascript source maps](https://developers.google.com/web/tools/chrome-devtools/javascript/source-maps). JavaParser could optionally look up for original source for regenerating untouched parts of code.

**Comment by davidsusu** [comment]
**Off.** I knew that there was a related thing from my online past, but I could not remember until now. I am not very active on SO, but about two years ago there was a question about text extracting from HTML. OP's problem was similar: search and extract from **HTML** by **text**. We can transform HTML to text (as a preprocessing), then search in the text for the wanted part, and finally: extract the original source of it. Also, I wrote a simplified example (in PHP). [My answer on SO](https://stackoverflow.com/questions/36690567/search-html-for-2-phrases-ignoring-all-tags-and-strip-everything-else/36694339#36694339).

**Comment by matozoid** [comment]
Hey @davidsusu, JavaParser keeps the output of the lexer around, you can ask every node what it's "token range" is and you will get the original source code (if any.) The issue here is that if we preprocess, we don't have a mapping to the source code anymore, but a mapping to the output of the preprocessor. Nobody will ever know what the original was. I think @cal101's solution to this same problem was to keep all information around, and do preprocessing, so that would be the better solution - at a cost.

Anyway, it is true that preprocessing is what the JLS prescribes. The solution implemented here is a simple but non-conformant solution to a common problem with JP, with the hopes that the part it breaks is so obscure that nobody will notice. So I'm really curious how you did notice, because this assumption seems wrong!

**Comment by matozoid** [comment]
@davidsusu - I'm still curious!

**Comment by davidsusu** [comment]
@matozoid Unit tests just fail. However I personally went over this problem: changed the tests.

I remember to only one production code where this would have been a problem, an old fashioned database mapping system with code generator, which created unicode escape sequences for non-english letters in field names. No longer in use. Legacy code can contain some kind of strange (but, of course, standard compliant strange).

**Comment by matozoid** [comment]
@davidsusu thanks - once you (or anyone else) runs into that kind of source I'll put a unicode preprocessor option in the next release.

**Comment by tibcoktredget** [comment]
I've just run into a problem with JavaParser 3.7.1 parsing some code generated by JavaCC:

`        jjtn000.setValue(t.image.substring(t.image.indexOf('\u005c'') + 1, t.image.length() - 1));
`

\u005c is the escape sequence for the backslash character, so this means:

`        jjtn000.setValue(t.image.substring(t.image.indexOf('\'') + 1, t.image.length() - 1));
`

but JavaParser does not parse this

Keith

**Comment by matozoid** [comment]
@tibcoktredget Interesting. Can you open this as a separate issue?

**Comment by tibcoktredget** [comment]
Thanks @matozoid for your prompt response - I've raised the issue

**Comment by davidsusu** [comment]
@tibcoktredget, @matozoid Isn't this the **same** problem? (The `char` literal parsed before JavaParser could recognize that the second single quote is escaped.)

**Comment by matozoid** [comment]
@davidsusu it's a process problem: these are replies to a closed PR. If a new change is needed, we need an open issue.

**Pull request merged**